lisa/tools/remote_copy.py

```diff
@@ -0,0 +1,85 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+import os
+from pathlib import PurePath
+
+from lisa.executable import Tool
+from lisa.tools.chown import Chown
+from lisa.tools.cp import Cp
+from lisa.tools.ls import Ls
+from lisa.tools.mkdir import Mkdir
+from lisa.tools.rm import Rm
+from lisa.tools.whoami import Whoami
+
+
+class RemoteCopy(Tool):
+    @property
+    def command(self) -> str:
+        return ""
+
+    @property
+    def can_install(self) -> bool:
+        return False
+
+    def _copy_dir(
+        self,
+        src: PurePath,
+        dest: PurePath,
+        recur: bool = True,
+    ) -> None:
+        # create the destination directory if it doesn't exist
+        destination_dir = dest / src.name
+        if not os.path.exists(destination_dir):
+            os.makedirs(destination_dir)
+
+        # get list of files and folders in the source directory
+        contents = self.node.tools[Ls].list(str(src))
+        dirs = [PurePath(content) for content in contents if content.endswith("/")]
+        files = [PurePath(content) for content in contents if not content.endswith("/")]
+
+        # copy files
+        for file in files:
+            self.node.shell.copy_back(file, destination_dir / file.name)
+
+        # return if recursion is not required
+        if not recur:
+            return
+
+        # copy folders
+        for dir in dirs:
+            self._copy_dir(dir, destination_dir, recur=recur)
+
+    def copy_dir(
+        self,
+        src: PurePath,
+        dest: PurePath,
+        recurse: bool = False,
+    ) -> None:
+        # if sudo is required, copy the files to a temp directory first
+        try:
+            # copy files to the destination directory
+            self._copy_dir(src, dest, recur=recurse)
+        except Exception as e:
+            self._log.debug(
+                f"Failed to copy files to {dest} with error: {e}, trying again"
+            )
+
+            # create a temp directory and copy the files to it
```

Comment:
this logic can be in a separated method like `_prepare_tmp_copy`, so the main logic is easier to understand.